xla/util.h

```diff
@@ -1018,6 +1018,45 @@ struct FreeDeleter {
 std::unique_ptr<void, FreeDeleter> AlignedAlloc(std::size_t alignment,
                                                 std::size_t size);
 
+// Serialized static variable initialization from the `func` output.
+// It avoid a deadlock between static variable initialization lock
+// and a lock in the `func` function.
+// Usage:
+// Instead of
+// static SomeType* p = [](){
+//     // for example we call some python code using nanobind
+//     nb::module_ numpy = nb::module_::import_("numpy");
+//     auto np_int8 = nb::object(numpy.attr("int8"));
+//     SomeType* obj = new SomeType(np_uint8);
+//     return obj;
+// };
+// We use SafeStaticInit
+// auto func = [](){
+//     // for example we call some python code using nanobind
+//     nb::module_ numpy = nb::module_::import_("numpy");
+//     auto np_int8 = nb::object(numpy.attr("int8"));
+//     SomeType* obj = new SomeType(np_uint8);
+//     return obj;
+// }
+// SomeType* p = SafeStaticInit<SomeType>(func);
+template<typename T, typename F>
+T * SafeStaticInit(F func) {
+  static absl::Mutex mutex;
+  static T* output = nullptr;
+  {
+    absl::MutexLock lock(&mutex);
+    if (output) {
+      return output;
+    }
+  }
+  T* p = func();
```

Comment:
This leaks a copy of T if someone else has already populated output. I would change func to return `std::unique_ptr<T>` and then use `output = p.release()` below?

Then if we end up not needing `p`, it will be freed.